### PR TITLE
[Feature] Support ImVoxelNet on KITTI

### PR DESCRIPTION
**Pull request by filaPro**
Reimplementation of ImVoxelNet for kitti-3d-car

### CODE REVIEW

File: configs/imvoxelnet/imvoxelnet_kitti-3d-car.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    neck_3d=dict(type='OutdoorImVoxelNeck', in_channels=64, out_channels=256),
+    bbox_head=dict(
+        type='Anchor3DHead',
+        num_classes=1,
+        in_channels=256,
+        feat_channels=256,
+        use_direction_classifier=True,
+        anchor_generator=dict(
+            type='Anchor3DRangeGenerator',
+            ranges=[[0, -39.68, -1.78, 69.12 - .32, 39.68 - .32, -1.78]],
```
Comment (Tai-Wang) [inline configs/imvoxelnet/imvoxelnet_kitti-3d-car.py:27]: - .32 seems a little strange? What is it used for here?

File: mmdet3d/models/detectors/imvoxelnet.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+from mmdet3d.core import bbox3d2result, build_anchor_generator
+from mmdet3d.models.fusion_layers.point_fusion import point_sample
+from mmdet.models import DETECTORS, build_backbone, build_head, build_neck
+from mmdet.models.detectors import BaseDetector
+
+
+@DETECTORS.register_module()
+class ImVoxelNet(BaseDetector):
+    """ImVoxelNet <https://arxiv.org/abs/2106.01178>."""
```
Comment (Tai-Wang) [inline mmdet3d/models/detectors/imvoxelnet.py:11]: r"""\`ImVoxelNet <https://arxiv.org/abs/2106.01178>\`_."""
(for the API doc automatic formatting)

File: mmdet3d/models/necks/imvoxel_neck.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        assert x.shape[-1] == 1
+        # Anchor3DHead axis order is (y, x).
+        return [x[..., 0].transpose(-1, -2)]
+
+    def init_weights(self):
+        """Initialize weights of neck."""
+        pass
+
+
+class ConvBlock(nn.Module):
```
Comment (Tai-Wang) [inline mmdet3d/models/necks/imvoxel_neck.py:44]: Can we use the ConvModule in the mmcv?

**Review by Tai-Wang — COMMENTED** [review]
Overall looks good to me. Just need to fix some minor comments.
In addition, maybe we can inherit the kitti base dataset config in the imvoxelnet config?

**Comment by filaPro** [comment]
> In addition, maybe we can inherit the kitti base dataset config in the imvoxelnet config?
I think there are too much point-cloud related transforms and we need only images.

Also, do I need to add a unit test?

**Comment by Tai-Wang** [comment]
> > In addition, maybe we can inherit the kitti base dataset config in the imvoxelnet config?
> 
> I think there are too much point-cloud related transforms and we need only images.
> 
> Also, do I need to add a unit test?

OK, then we can keep the current format for configs.
Unit tests are needed, we'd better also add them in this PR.
I observe that the linting failed. You can refer to this [guidance](https://github.com/open-mmlab/mmdetection3d/blob/master/.github/CONTRIBUTING.md) for useful pre-commit tools to avoid this problem.

File: mmdet3d/models/detectors/imvoxelnet.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        self.backbone.init_weights(pretrained=pretrained)
+        self.neck.init_weights()
+        self.neck_3d.init_weights()
+        self.bbox_head.init_weights()
+
+    def extract_feat(self, img, img_metas):
+        """Extract 3d features from the backbone -> fpn -> 3d projection.
+
+        Args:
+            img (Tensor): Input images of shape (N, C_in, H, W).
```
Comment (Tai-Wang) [inline mmdet3d/models/detectors/imvoxelnet.py:53]: Tensor -> torch.Tensor, please also modify others to be consistent

**Review by Tai-Wang — APPROVED** [review]
Almost ready to merge except consistent expression of Tensor in the docstring

File: configs/imvoxelnet/imvoxelnet_kitti-3d-car.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+total_epochs = 12
+
+checkpoint_config = dict(interval=1, max_keep_ckpts=1)
+log_config = dict(
+    interval=50,
+    hooks=[dict(type='TextLoggerHook'),
+           dict(type='TensorboardLoggerHook')])
+evaluation = dict(interval=1)
+dist_params = dict(backend='nccl')
+find_unused_parameters = True  # only 1 of 4 FPN outputs is used
```
[the remark this answers is not in the document]
Comment (filaPro) [reply configs/imvoxelnet/imvoxelnet_kitti-3d-car.py:156]: Probably it is just a bad design choice. For some reason we followed the implementation in [Atlas](https://github.com/magicleap/Atlas).

**Comment by filaPro** [comment]
I will also run training on 8 GPUs one more time to be sure that refactoring Anchor3DGenerator and ConvModule not much influenced anything.

File: configs/imvoxelnet/imvoxelnet_kitti-3d-car.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Tai-Wang) [reply configs/imvoxelnet/imvoxelnet_kitti-3d-car.py:156]: OK, then let's keep it at first. Maybe we can study more choices here in the future.

**Comment by filaPro** [comment]
I reverted anchor ranges, metrics are fine now. Can share model and log if needed.

**Review by Tai-Wang — APPROVED** [review]

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**